src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -908,33 +921,61 @@ protected long logEvent(String eventType, JSONObject eventProperties, JSONObject
             event.put("user_id", replaceWithJSONNull(userId));
             event.put("device_id", replaceWithJSONNull(deviceId));
             event.put("session_id", outOfSession ? -1 : sessionId);
-            event.put("version_name", replaceWithJSONNull(deviceInfo.getVersionName()));
-            event.put("os_name", replaceWithJSONNull(deviceInfo.getOsName()));
-            event.put("os_version", replaceWithJSONNull(deviceInfo.getOsVersion()));
-            event.put("device_brand", replaceWithJSONNull(deviceInfo.getBrand()));
-            event.put("device_manufacturer", replaceWithJSONNull(deviceInfo.getManufacturer()));
-            event.put("device_model", replaceWithJSONNull(deviceInfo.getModel()));
-            event.put("carrier", replaceWithJSONNull(deviceInfo.getCarrier()));
-            event.put("country", replaceWithJSONNull(deviceInfo.getCountry()));
-            event.put("language", replaceWithJSONNull(deviceInfo.getLanguage()));
-            event.put("platform", platform);
             event.put("uuid", UUID.randomUUID().toString());
             event.put("sequence_number", getNextSequenceNumber());
 
+            if (trackingOptions.shouldTrackVersionName()) {
+                event.put("version_name", replaceWithJSONNull(deviceInfo.getVersionName()));
+            }
+            if (trackingOptions.shouldTrackOsName()) {
+                event.put("os_name", replaceWithJSONNull(deviceInfo.getOsName()));
+            }
+            if (trackingOptions.shouldTrackOsVersion()) {
+                event.put("os_version", replaceWithJSONNull(deviceInfo.getOsVersion()));
+            }
+            if (trackingOptions.shouldTrackDeviceBrand()) {
+                event.put("device_brand", replaceWithJSONNull(deviceInfo.getBrand()));
+            }
+            if (trackingOptions.shouldTrackDeviceManufacturer()) {
+                event.put("device_manufacturer", replaceWithJSONNull(deviceInfo.getManufacturer()));
+            }
+            if (trackingOptions.shouldTrackDeviceModel()) {
```

Comment:
Unclear how much this affects performance, basically for each of these fields we need to do a Set contains check. For each field, for each event.